src/v/compat/run.cc

```diff
@@ -0,0 +1,196 @@
+            w.EndObject();
+        }
+        w.EndArray();
+        w.EndObject();
+    }
+
+    /*
+     * Writes all test cases to the provided directory.
+     */
+    static ss::future<> write(const std::filesystem::path& dir) {
+        size_t instance = 0;
+
+        for (auto& test : checker::create_test_cases()) {
+            // json encoded test case
+            auto buf = json::StringBuffer{};
+            auto writer = json::Writer<json::StringBuffer>{buf};
+            write_test_case(std::move(test), writer);
+
+            // save test case to file
+            iobuf data;
+            data.append(buf.GetString(), buf.GetSize());
+            auto fn = fmt::format("{}_{}.json", checker::name, instance++);
+            write_fully(dir / fn, std::move(data)).get();
+        }
+
+        return ss::now();
+    }
+
+    /*
+     * Check a test case.
+     */
+    static void check(json::Document doc) {
+        // fields is the content of the test instance
+        vassert(doc.HasMember("fields"), "document doesn't contain fields");
+        vassert(doc["fields"].IsObject(), "fields is not an object");
+        auto instance = checker::from_json(doc["fields"].GetObject());
+
+        // binary encodings of the test instance
+        vassert(doc.HasMember("binaries"), "document doesn't contain binaries");
+        vassert(doc["binaries"].IsArray(), "binaries is not an array");
+        auto binaries = doc["binaries"].GetArray();
+
+        for (const auto& encoding : binaries) {
+            vassert(encoding.IsObject(), "binaries entry is not an object");
+            auto binary = encoding.GetObject();
+
+            vassert(binary.HasMember("name"), "encoding doesn't have name");
+            vassert(binary["name"].IsString(), "encoding name is not string");
+            vassert(binary.HasMember("data"), "encoding doesn't have data");
+            vassert(binary["data"].IsString(), "encoding data is not string");
+
+            compat_binary data(
+              binary["name"].GetString(),
+              bytes_to_iobuf(base64_to_bytes(binary["data"].GetString())));
+
+            // keep copy for next round
+            auto&& [ia, ib] = compat_copy(std::move(instance));
+            instance = std::move(ia);
+
+            if (!checker::check(std::move(ib), std::move(data))) {
+                json::StringBuffer buf;
+                json::PrettyWriter<json::StringBuffer> writer(buf);
+                doc.Accept(writer);
+                fmt::print("Input JSON: {}\n", buf.GetString());
+                throw compat_error(
+                  fmt::format("failed checking {{{}}}\n", checker::name));
+            }
+        }
+    }
+};
+
+template<typename T>
+static void maybe_check(std::string name, json::Document& doc, bool& seen) {
+    using type = corpus_helper<T>;
+    if (type::checker::name == name) {
+        vassert(!seen, "Duplicate check found: {}", name);
+        seen = true;
```

Comment:
hmm this presents an interesting problem. However maybe more static type magic can help us here. If we can assert that there can only be uniquely named types within the typelist, then we can guarantee that a duplicate check would never exist. Maybe to make that simpler a one time verification step can be done on the static typelist, something like:
```
template<typename... Types>
void verify(type_list<Types...> ts) {
    const auto arr = std::to_array({Types::name...}); 
    const std::set<std::string> set = {Types::name...};
    vassert (set.size() == arr.size(), "Duplicate detected"); 
```
This could probably be improved if the method would be `constexpr` then you'd have this verification at compile time.

Reply:
So I got this to work only on gcc, apparently there is a clang bug about comparison of constant expressions, https://godbolt.org/z/vvqPsr83M

Reply:
nice. this seems to work